### ISSUE TITLE
Is this intersection correct?

### ISSUE THREAD

**Issue by pauljurczak** [ajhynes7/scikit-spatial#283]
This code snippet:
```
cylinder = Cylinder([0, 0, 0], [0, 0, 1], 1)
line = Line([2, 2, 2], [3, 3, 3])
print(cylinder.intersect_line(line))
```
produces:
```
(Point([-0.70710678, -0.70710678, -0.70710678]), Point([0.70710678, 0.70710678, 0.70710678]))
```
This doesn't seem to be correct. The line intersects the cylinder at [0.70710678, 0.70710678, 0.70710678] and [0, 0, 0]:

![cylinder](https://user-images.githubusercontent.com/4933032/134768110-15a76d35-2c5c-45ad-90df-926db0d4a643.png)

**Comment by ajhynes7** [comment]
The method `intersect_line` on class `Cylinder` treats the cylinder as having no caps. The following is from the docstring:

> This method treats the cylinder as infinite along its axis (i.e., without caps).

That's why you don't get the intersection point `[0, 0, 0]`. If you're interested, I can try writing a method that calculates the intersection points keeping the caps in account. It'll be a bit more complicated than the current method.

**Comment by pauljurczak** [comment]
Sorry, I must have glossed over the infinite cylinder part. Yes, I'm definitely interested in finite length cylinder intersection. Specifically, I just need a boolean function: intersects or doesn't intersect. What would be the performance? I have to call it about 100,000,000 times.

**Comment by ajhynes7** [comment]
I have an idea of how it could be written. If you're interested in performance, would you be willing to try Julia? I've started a Julia implementation of this package called [ScikitSpatial.jl](https://github.com/ajhynes7/ScikitSpatial.jl). It's not as feature-complete as this package, but I think it would provide better performance.

**Comment by pauljurczak** [comment]
I've played a bit with Julia, but can't use it for this project. I will adapt the infinite cylinder for now, since I'm dealing with a finite number of its segments. I found that ray casting is really the most appropriate for my use case. That would require defining a new spatial object: Ray. Only the first surface intersection would be of interest.

**Comment by pauljurczak** [comment]
BTW, minor issue, but the cylinder is plotted as a finite shape:
```
cylinder = Cylinder([0, 0, 0], [0, 0, 1], 1)
line = Line([2, 2, 2], [3, 3, 3])
points = Points(cylinder.intersect_line(line))

plot_3d(
  line.plotter(t_1=-1, c='blue', linestyle='--'),
  cylinder.plotter(color='red'),
  points.plotter(color='green')
)

plt.show()
```
produces:

![cyl](https://user-images.githubusercontent.com/4933032/134828893-86fe75ed-bd40-4948-a596-f40c0d7db9b0.png)

which looks a bit misleading.

**Comment by ajhynes7** [comment]
The cylinder object itself is finite, which is why it has a `length` method. Only the `intersect_line` method uses an infinite cylinder to compute the intersection points, because it was a simpler computation.

**Comment by pauljurczak** [comment]
I see. I made too many assumptions…

**Comment by pauljurczak** [comment]
Perhaps renaming `intersect_line` to `extend_and_intersect_line` or something similar would alert the user about cylinder being extended to infinity before computing the intersection.

**Comment by ajhynes7** [comment]
I'm planning to add an `infinite=True` parameter to the `intersect_line` method, so it's not a breaking change. I'll need to write the code for the non-infinite cylinder.

Do you still need the code for checking if a line intersects with the cylinder? I have something locally that I think works.

**Comment by pauljurczak** [comment]
True, `infinite=True` makes more sense than changing the API. No rush on my side. I may come back to experimenting with scikit-spatial, but for now I'm going to use trimesh (https://github.com/mikedh/trimesh), which has a very fast implementation based on triangular meshes. I don't need high accuracy for my project.
